Attack algorithms

BrianDunlap89 opened the pull request:
@CodingItWrong moved random attack logic `character.rb` into the new `random_attack.rb` class. Also updated the feature specs to account for the change.

BrianDunlap89 committed:
Move direct damage logic from the character class into a distinct simple attack class

BrianDunlap89 committed:
Simplify receive_damage method to calculate only the character's current health

BrianDunlap89 committed:
Create random attack class to calculate damage with a variable multiplier

BrianDunlap89 committed:
Require new attack classes in battle_logic.rb file

BrianDunlap89 committed:
Substitute all character method calls for 'attack' with the calls to instantiations of the appropriate attack class

CodingItWrong commented on lib/battle_logic/random_attack.rb:
```diff
@@ -0,0 +1,35 @@
+  attr_reader :attacker, :defender
+
+  def perform
+    defender.receive_damage!(damage_after_defense)
```
With the way this class has come together, I think we might be able to represent `damage_after_defense` as just `damage`. It's the net damage for the attack.

CodingItWrong commented on lib/battle_logic/random_attack.rb:
```diff
@@ -0,0 +1,35 @@
+  private
+
+  def damage_after_defense
+    (attack_damage * defense_modifier) + scratch_damage
```
What do you think about the pros and cons of inlining `attack_damage` here? Now that the attack modifier and defense modifier calculation both happen in the same class, I'm not sure whether or not naming `attack_damage` as a concept helps us. But `defense_after_damage` does start to get a bit long if it gets inlined… Not sure.

CodingItWrong commented on lib/battle_logic/random_attack.rb:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  def defense_modifier
+    (256.0 - defender.defense_rating) / 256.0
```
I bet `256.0` can be a named concept. Maybe a constant?

CodingItWrong commented on lib/battle_logic/simple_attack.rb:
```diff
@@ -0,0 +1,23 @@
+
+  def damage_after_defense
+    damage = attacker.attack_rating
+    min_zero(damage - defender.defense_rating)
```
This one feels like if it got combined into one line it could be nice. `attacker.attack_rating - defender.defense_rating` reads really simply.

CodingItWrong commented on spec/features/performing_attacks_spec.rb:
```diff
@@ -1,88 +1,96 @@
+      it "allows simple characters to be killed in one hit" do
+        mario = BattleLogic::Character.new
+        goomba = BattleLogic::Character.new
+        SimpleAttack.new(attacker: mario, defender: goomba).perform
```
I think I might play around different ways to write instantiating the attack. Having it wrapped up in a class is definitely the right thing to do, but it'd be cool if it read more like English. Not sure. Calling it on `mario` before was nice—maybe `Character` can have a really simple `attack` method that delegates through to a configured `Attack` class somehow.

BrianDunlap89 replied:
Agreed. It felt weird to have the attack object "do" the attacking.

CodingItWrong commented on spec/features/performing_attacks_spec.rb:
```diff
@@ -1,88 +1,96 @@
-      expect(bowser).to be_alive
-    end
+      it "does not kill the character after two hits" do
+        2.times { attack.perform }
```
I forgot about the fact that instantiating and calling `perform` separately makes it easy to reuse an attack like this. Tests driving design!

CodingItWrong replied:
Actually I didn't forget; you changed it to work this way. I like it a lot!

CodingItWrong commented on spec/features/performing_attacks_spec.rb:
```diff
@@ -1,88 +1,96 @@
+    context "multiple points of damage" do
+      let(:mario) { BattleLogic::Character.new(attack_rating: 2) }
+      let(:bowser) { BattleLogic::Character.new(max_health: 3) }
+      let(:attack) { SimpleAttack.new(attacker: mario, defender: bowser) }
```
Could be cool to pull `SimpleAttack` here (and `RandomAttack` later) out into a `let()`, so you're kind of defining at the top of the `context` what it is you're specifying there. Or maybe that means they should be different feature specs.

BrianDunlap89 replied:
Ha, yeah I think the reason I didn't define a `SimpleAttack` for the entire context is because you used the goomba/mario combination for the first example, and I liked the goomba reference. They're an underrepresented species! But yes, I think you're right.

BrianDunlap89 replied:
As for creating a separate feature spec for each attack type, it seems a little superfluous now, but would probably be helpful if our algorithms / attack types became more complex/nuanced. What do you recommend? Proactively compartmentalizing things? Or only doing so if it reaches a threshold of complexity?

CodingItWrong replied:
Haha, gotcha. I didn't want to specify koopa shell behavior at this point.

This starts to get into what exactly are we specifying in this file. When there was only a `Character` class, it was all in this file. Now that there are `Attack` classes too… We may not even need to specify the `Character` behavior at the feature level anymore. 🤔

CodingItWrong replied:
I recommend proactively compartmentalizing things, but the TDD approach I'm trying to learn recommends waiting, so let's go with waiting! I trust them way more than I trust my background.

CodingItWrong commented on spec/features/performing_attacks_spec.rb:
```diff
@@ -1,88 +1,96 @@
     context "outleveled" do
       let(:terra) { BattleLogic::Character.new(attack_rating: 150) }
-      let(:kefka) { BattleLogic::Character.new(defense_rating: 255, max_health: 9999) }
+      let(:kefka) { BattleLogic::Character.new(defense_rating: 256, max_health: 9999) }
```
Nice fix of the numbers to get the intent working. I'm sure you could tell from how I was talking that I hadn't thought through off-by-one issues in depth when I defined the test and the pseudocode algorithm.

CodingItWrong commented:
@BrianDunlap89 looks good! could be cool to unit-test the attack classes too. I had been thinking about the fact that a unit test for them wouldn't really be mocking out `Character`s per se, but rather a kind of role or duck type: just whatever method calls these attacks need them to have.

Do you want to do anything else on this PR? (1) nothing, (2) just the suggestions above that feel like "minor" changes, or (3) some of the more substantial ideas I had (or that you have)?

Commit:
Get all tests passing and re-adds attack method

Commit:
Always require spec helper

Commit:
Move instantiation into a factory

This allows for consistent character instance configuration, i.e. always using the same attack action

Commit:
Namespace attacks into BL module

Commit:
Specify attack actions

Commit:
Rename models to units

CodingItWrong commented:
@BrianDunlap89 take a look at what i ended up doing when you have a chance. major things:
- I re-introduced `Character#attack`, and had the class of attack be configured on the character
- But then that felt a bit hairy, so I moved character instantiation into a Factory class, and that Factory class ensures that all characters are configured with the same type of attack.
- I unit tested the attacks, and it felt very clean to test the logic in there directly, and with no dependencies on the class that attacker and defender are
- As Sandi Metz describes in her testing chapter (but using Minitest instead of RSpec), I described roles/duck types with shared examples, and then included those in the appropriate unit tests. Very readable to see `it_behaves_like "an action"`.
